daf_tests sometimes fails with panic: interface conversion: error is *url.Error, not *main.AppError

## Expected Behavior

daf_tests should complete normally

## Actual Behavior

It appears that the daf_tests is not gracefully handling errors when the patients app is not responding

## Steps to Reproduce the Problem

Stop the application on testweb and then run daf_tests

## Version

Indicate the relevant versions of components. This may include the Titan
package version, operating system version, or other components relevant to
the project.

## Additional Information

```
+ /var/lib/jenkins/daf_tests

2019/11/01 21:01:40 Initializing Resty Client

2019/11/01 21:01:40 Logging in to http://testweb:8080/auth/login

RESTY 2019/11/01 21:01:40 ERROR [Post http://testweb:8080/auth/login: dial tcp 10.0.1.73:8080: connect: connection refused] Attempt [1]

RESTY 2019/11/01 21:01:45 ERROR [Post http://testweb:8080/auth/login: dial tcp 10.0.1.73:8080: connect: connection refused] Attempt [2]

RESTY 2019/11/01 21:01:53 ERROR [Post http://testweb:8080/auth/login: dial tcp 10.0.1.73:8080: connect: connection refused] Attempt [3]

2019/11/01 21:02:05 Fetching all patients

2019/11/01 21:02:05 http://testweb:8080/patients/

RESTY 2019/11/01 21:02:05 ERROR [Get http://testweb:8080/patients/: dial tcp 10.0.1.73:8080: connect: connection refused] Attempt [1]

RESTY 2019/11/01 21:02:10 ERROR [Get http://testweb:8080/patients/: dial tcp 10.0.1.73:8080: connect: connection refused] Attempt [2]

RESTY 2019/11/01 21:02:16 ERROR [Get http://testweb:8080/patients/: dial tcp 10.0.1.73:8080: connect: connection refused] Attempt [3]

panic: interface conversion: error is *url.Error, not *main.AppError



goroutine 1 [running]:

main.main()

	/Users/abowen/Downloads/scripts/daf_tests/main.go:25 +0x1c4f

script returned exit code 2
```

--- a/ansible/roles/tooling_oracle_config/tasks/main.yml
+++ b/ansible/roles/tooling_oracle_config/tasks/main.yml
@@ -69,9 +69,9 @@
   tags:
     - delphixpy
 
-- name: Copy daf_tests
+- name: Create daf_tests
   copy: 
-    src: daf_tests
+    src: ../../../../go/daf_tests/bin/linux64/daf_tests
     dest: "/var/lib/jenkins/daf_tests"
     owner: jenkins
     group: jenkins

--- a/demo-workshops/tcw/docs/release_notes.md
+++ b/demo-workshops/tcw/docs/release_notes.md
@@ -2,6 +2,12 @@
 
 ## What's Changed
 
+### v2.14
+
+1. (143bd6e) Jenkins uses Java 11
+2. (d17999b) Jenkins plugins pinned (#32)
+3. (1ea60b1) Fixed #33
+
 ### v2.13
 
 1. (50989ae) #20 Added Delphix Jenkins plugin

--- a/demo-workshops/tcw/workshop.yaml
+++ b/demo-workshops/tcw/workshop.yaml
@@ -48,7 +48,7 @@ amis:
 - delphix-tcw-tooling-oracle:
     packer:
       batch: 4
-      md5sum: 018ac47f8040e9ba0fc48cc33145d2c0
+      md5sum: f693f36cb0f6a7fbb2e5fae5b52188ba
       name: delphix-tcw-tooling-oracle.json
 - delphix-tcw-virtualizationengine: {}
 - delphix-tcw-maskingengine: {}

--- a/go/daf_tests/.gitignore
+++ b/go/daf_tests/.gitignore
@@ -1,0 +1,2 @@
+build
+bin

--- a/go/daf_tests/Makefile
+++ b/go/daf_tests/Makefile
@@ -1,0 +1,16 @@
+# Makefile
+
+.PHONY: build
+
+local: 
+	go run *.go
+
+local-debug:
+	go run *.go -vvv
+
+build:
+	echo $$GOPATH
+	go get -d
+	env CGO_ENABLED=0 GOOS=linux GOARCH=amd64 go build -i -v -ldflags="-X main.version=1.0.2" -o ./bin/linux64/daf_tests
+
+	

--- a/go/daf_tests/billings.go
+++ b/go/daf_tests/billings.go
@@ -1,0 +1,35 @@
+package main
+
+import (
+	"fmt"
+	"log"
+
+	resty "gopkg.in/resty.v1"
+)
+
+func (c *Client) getBillingsFromAnyPatientInList(patientList map[string]interface{}) (map[string]interface{}, error) {
+	patientID, err := getPseudoRandomPatientIDFromPatientsList(patientList)
+	if err != nil {
+		return nil, err
+	}
+	return c.getBillings(patientID)
+}
+
+func (c *Client) getBillings(patient int) (map[string]interface{}, error) {
+
+	url := fmt.Sprintf("%s/patients/%d/billings", c.url, patient)
+	log.Printf("Fetching all billings for patient %d", patient)
+	log.Println(url)
+	resp, err := resty.R().
+		Get(url)
+	if err != nil {
+		return nil, err
+	}
+
+	bodyMap, err := parseHTTPResponseReturnMap(resp)
+	if err != nil {
+		return nil, err
+	}
+
+	return bodyMap, nil
+}

--- a/go/daf_tests/config.go
+++ b/go/daf_tests/config.go
@@ -1,0 +1,116 @@
+package main
+
+import (
+	"encoding/json"
+	"fmt"
+	"log"
+	"net/http"
+	"time"
+
+	resty "gopkg.in/resty.v1"
+)
+
+func init() {
+	log.Printf("Version: %v", version)
+}
+
+type Client struct {
+	url, username, password string
+}
+
+type AppError struct {
+	message, timestamp, errType string
+	status                      int
+}
+
+func newAppError(resp *resty.Response) *AppError {
+	var errMap map[string]interface{}
+	result := resp.Body()
+	if err := json.Unmarshal(result, &errMap); err != nil {
+		log.Fatal(err)
+	}
+	log.Printf("%v", errMap)
+	return &AppError{
+		message:   errMap["message"].(string),
+		timestamp: errMap["timestamp"].(string),
+		errType:   errMap["error"].(string),
+		status:    int(errMap["status"].(float64)),
+	}
+}
+
+func (e *AppError) Error() string {
+	return e.message
+}
+
+// NewClient creates a new client object
+func NewClient(username, password, url string) *Client {
+	return &Client{
+		url:      url,
+		username: username,
+		password: password,
+	}
+}
+
+func (c *Client) initResty() {
+	log.Printf("Initializing Resty Client")
+	resty.DefaultClient.
+		SetTimeout(time.Duration(30 * time.Second)).
+		SetRetryCount(3).
+		SetRetryWaitTime(5 * time.Second).
+		SetRetryMaxWaitTime(20 * time.Second)
+
+	resty.
+		SetHeader("Content-Type", "application/json").
+		SetHeader("Accept", "application/json").
+		SetHeader("User-Agent", "PostmanRuntime/7.13.0").
+		SetHeader("Accept", "*/*").
+		SetHeader("Cache-Control", "no-cache").
+		SetHeader("accept-encoding", "gzip, deflate").
+		SetHeader("Connection", "keep-alive").
+		SetHeader("cache-control", "no-cache")
+}
+
+func parseHTTPResponse(resp *resty.Response) ([]byte, error) {
+	var err error
+
+	if http.StatusOK != resp.StatusCode() {
+		err = newAppError(resp)
+		if err != nil {
+			return nil, err
+		}
+	}
+	return resp.Body(), err
+}
+
+func parseHTTPResponseReturnSlice(resp *resty.Response) (resultdat []interface{}, err error) {
+	result, err := parseHTTPResponse(resp)
+	if err != nil {
+		return nil, err
+	}
+	if err := json.Unmarshal(result, &resultdat); err != nil { //convert the json to go objects
+		return nil, err
+	}
+
+	return resultdat, nil
+}
+
+func parseHTTPResponseReturnMap(resp *resty.Response) (resultdat map[string]interface{}, err error) {
+	result, err := parseHTTPResponse(resp)
+	if err != nil {
+		return nil, err
+	}
+	if err = json.Unmarshal(result, &resultdat); err != nil { //convert the json to go objects
+		return nil, err
+	}
+
+	return resultdat, err
+}
+
+func getPseudoRandomPatientIDFromPatientsList(patientList map[string]interface{}) (int, error) {
+	patientID, ok := patientList["content"].([]interface{})[0].(map[string]interface{})["id"].(float64)
+	if !ok {
+		err := fmt.Errorf("Did not find a patientID. Something went terribly wrong")
+		return 0, err
+	}
+	return int(patientID), nil
+}

--- a/go/daf_tests/junitxml.go
+++ b/go/daf_tests/junitxml.go
@@ -1,0 +1,85 @@
+package main
+
+import (
+	"encoding/xml"
+	"fmt"
+	"io"
+	"log"
+	"os"
+)
+
+type TestSuite struct {
+	XMLName    xml.Name `xml:"testsuite"`
+	Errors     string   `xml:"errors,attr"`
+	Failures   string   `xml:"failures,attr"`
+	Name       string   `xml:"name,attr"`
+	Tests      int      `xml:"tests,attr"`
+	Properties *Properties
+	TestCases  []TestCase
+}
+
+type Properties struct {
+	XMLName       xml.Name `xml:"properties"`
+	PropertyArray []*Property
+}
+
+type Property struct {
+	XMLName xml.Name `xml:"property"`
+	Name    string   `xml:"name,attr"`
+	Value   string   `xml:"value,attr"`
+}
+
+type TestCase struct {
+	XMLName     xml.Name `xml:"testcase"`
+	Classname   string   `xml:"classname,attr"`
+	Name        string   `xml:"name,attr"`
+	TestError   *TestError
+	TestFailure *TestFailure
+	TestSuccess *TestSuccess
+}
+
+type TestError struct {
+	XMLName   xml.Name `xml:"error"`
+	Type      string   `xml:"type,attr,omitempty"`
+	Timestamp string   `xml:"timestamp,attr,omitempty"`
+	Status    int      `xml:"status,attr,omitempty"`
+	Message   string   `xml:",chardata"`
+}
+
+type TestFailure struct {
+	XMLName   xml.Name `xml:"failure"`
+	Type      string   `xml:"type,attr,omitempty"`
+	Timestamp string   `xml:"timestamp,attr,omitempty"`
+	Status    int      `xml:"status,attr,omitempty"`
+	Message   string   `xml:",chardata"`
+}
+
+type TestSuccess struct {
+	XMLName   xml.Name `xml:"success"`
+	Type      string   `xml:"type,attr,omitempty"`
+	Timestamp string   `xml:"timestamp,attr,omitempty"`
+	Status    int      `xml:"status,attr,omitempty"`
+	Message   string   `xml:",chardata"`
+}
+
+func createReportFile(path, filename string) *os.File {
+	err := os.MkdirAll(path, 0744)
+	if err != nil {
+		log.Fatalf("error creating path: %v", err)
+	}
+	file, err := os.OpenFile(fmt.Sprintf("%s/%s", path, filename), os.O_TRUNC|os.O_CREATE|os.O_WRONLY, 0644)
+	if err != nil {
+		log.Fatalf("error opening file: %v", err)
+	}
+	return file
+}
+
+func writeReport(report *os.File, ts *TestSuite) {
+	xmlWriter := io.Writer(report)
+	xmlWriter.Write([]byte(xml.Header))
+	enc := xml.NewEncoder(xmlWriter)
+	enc.Indent("  ", "    ")
+	if err := enc.Encode(ts); err != nil {
+		log.Fatalf("Unable to write report: %v\n", err)
+	}
+}

--- a/go/daf_tests/login.go
+++ b/go/daf_tests/login.go
@@ -1,0 +1,50 @@
+package main
+
+import (
+	"fmt"
+	"log"
+	"net/http"
+
+	resty "gopkg.in/resty.v1"
+)
+
+func (c *Client) loginTest() error {
+	log.Printf("Testing login")
+	err := c.login()
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (c *Client) login() error {
+	c.initResty()
+	url := c.url + "/auth/login"
+	log.Printf("Logging in to %s", url)
+
+	payload := fmt.Sprintf("{\n\t\"username\": \"%s\",\n\t\"password\": \"%s\"\n}", c.username, c.password)
+
+	resp, err := resty.R().
+		SetBody(payload).
+		Post(url)
+	if err != nil {
+		return err
+	}
+
+	if http.StatusUnauthorized == resp.StatusCode() {
+		err = fmt.Errorf("%d", http.StatusUnauthorized)
+		if err != nil {
+			return err
+		}
+	}
+
+	_, err = parseHTTPResponse(resp)
+	if err != nil {
+		return err
+	}
+
+	resty.
+		SetHeader("Authorization", fmt.Sprintf("Bearer %s", resp.Body()))
+
+	return nil
+}

--- a/go/daf_tests/main.go
+++ b/go/daf_tests/main.go
@@ -1,0 +1,144 @@
+package main
+
+import (
+	"fmt"
+	"log"
+)
+
+var (
+	version = "not set"
+)
+
+func main() {
+	testFail := false
+	v := &TestSuite{Name: "daf_tests"}
+	v.Properties = &Properties{}
+	v.Properties.PropertyArray = append(v.Properties.PropertyArray, &Property{Name: "tests-group", Value: "daf_tests"})
+	report := createReportFile("build/reports", "test.xml")
+	defer report.Close()
+
+	server := "testweb"
+	port := "8080"
+	username := "patients_admin"
+	password := "delphix"
+	client := NewClient(username, password, fmt.Sprintf("http://%s:%s", server, port))
+	err := client.login()
+	if err != nil {
+		log.Fatal(err)
+	}
+	testCase := TestCase{Classname: "Read All Patients"}
+	patientList, err := client.getPatients()
+	if err != nil {
+		appError := err.(*AppError)
+		testCase.TestFailure = &TestFailure{
+			Type:      appError.errType,
+			Timestamp: appError.timestamp,
+			Status:    appError.status,
+			Message:   appError.message,
+		}
+		testFail = true
+	} else {
+		testCase.TestSuccess = &TestSuccess{Message: fmt.Sprintf("%v", patientList)}
+	}
+	v.TestCases = append(v.TestCases, testCase)
+
+	testCase = TestCase{Classname: "Read All Records"}
+	records, err := client.getRecordsFromAnyPatientInList(patientList)
+	if err != nil {
+		appError := err.(*AppError)
+		testCase.TestFailure = &TestFailure{
+			Type:      appError.errType,
+			Timestamp: appError.timestamp,
+			Status:    appError.status,
+			Message:   appError.message,
+		}
+		testFail = true
+	} else {
+		testCase.TestSuccess = &TestSuccess{Message: fmt.Sprintf("%v", records)}
+	}
+	v.TestCases = append(v.TestCases, testCase)
+
+	testCase = TestCase{Classname: "Read All Billings"}
+	billings, err := client.getBillingsFromAnyPatientInList(patientList)
+	if err != nil {
+		appError := err.(*AppError)
+		testCase.TestFailure = &TestFailure{
+			Type:      appError.errType,
+			Timestamp: appError.timestamp,
+			Status:    appError.status,
+			Message:   appError.message,
+		}
+		testFail = true
+	} else {
+		testCase.TestSuccess = &TestSuccess{Message: fmt.Sprintf("%v", billings)}
+	}
+	v.TestCases = append(v.TestCases, testCase)
+
+	testCase = TestCase{Classname: "Read All Payments"}
+	payments, err := client.getPaymentsFromAnyPatientInList(patientList)
+	if err != nil {
+		appError := err.(*AppError)
+		testCase.TestFailure = &TestFailure{
+			Type:      appError.errType,
+			Timestamp: appError.timestamp,
+			Status:    appError.status,
+			Message:   appError.message,
+		}
+		testFail = true
+	} else {
+		testCase.TestSuccess = &TestSuccess{Message: fmt.Sprintf("%v", payments)}
+	}
+	v.TestCases = append(v.TestCases, testCase)
+
+	testCase = TestCase{Classname: "Add/Delete User"}
+	bob := &User{"bobbyb", "bob", "bee", "beebee"}
+	userID, err := client.signUpUserAndReturnID(bob)
+
+	log.Printf("User %s assigned ID %d", bob.username, userID)
+	testClient := NewClient(bob.username, bob.password, client.url)
+	err = testClient.loginTest()
+	if err != nil {
+		appError := err.(*AppError)
+		testCase.TestFailure = &TestFailure{
+			Type:      appError.errType,
+			Timestamp: appError.timestamp,
+			Status:    appError.status,
+			Message:   appError.message,
+		}
+		testFail = true
+	}
+	client.login()
+	err = client.deleteUser(userID)
+	if err != nil {
+		appError := err.(*AppError)
+		testCase.TestFailure = &TestFailure{
+			Type:      appError.errType,
+			Timestamp: appError.timestamp,
+			Status:    appError.status,
+			Message:   appError.message,
+		}
+		testFail = true
+	}
+	v.TestCases = append(v.TestCases, testCase)
+
+	testCase = TestCase{Classname: "Add Duplicate User"}
+	michael := &User{"mcred", "Michael", "Credible", "Ibreakthings"}
+	userID, err = client.signUpUserAndReturnID(michael)
+	if err != nil {
+		testCase.TestSuccess = &TestSuccess{Message: fmt.Sprintf("%v", err)}
+	} else {
+		testCase.TestFailure = &TestFailure{
+			Type:    "Unique Constraint",
+			Status:  200,
+			Message: fmt.Sprintf("User %s assigned ID %d", michael.username, userID),
+		}
+		testFail = true
+	}
+
+	v.TestCases = append(v.TestCases, testCase)
+	writeReport(report, v)
+	if testFail == true {
+		// os.Exit(2)
+		log.Print("tests failed")
+	}
+}

--- a/go/daf_tests/patients.go
+++ b/go/daf_tests/patients.go
@@ -1,0 +1,34 @@
+package main
+
+import (
+	"fmt"
+	"log"
+	"net/http"
+
+	resty "gopkg.in/resty.v1"
+)
+
+func (c *Client) getPatients() (map[string]interface{}, error) {
+
+	url := c.url + "/patients/"
+	log.Println("Fetching all patients")
+	log.Println(url)
+	resp, err := resty.R().
+		Get(url)
+	if err != nil {
+		return nil, err
+	}
+
+	if http.StatusUnauthorized == resp.StatusCode() {
+		err = fmt.Errorf("%d", http.StatusUnauthorized)
+		if err != nil {
+			return nil, err
+		}
+	}
+	bodyMap, err := parseHTTPResponseReturnMap(resp)
+	if err != nil {
+		return nil, err
+	}
+
+	return bodyMap, nil
+}

--- a/go/daf_tests/payments.go
+++ b/go/daf_tests/payments.go
@@ -1,0 +1,36 @@
+package main
+
+import (
+	"fmt"
+	"log"
+
+	resty "gopkg.in/resty.v1"
+)
+
+func (c *Client) getPaymentsFromAnyPatientInList(patientList map[string]interface{}) (map[string]interface{}, error) {
+	patientID, err := getPseudoRandomPatientIDFromPatientsList(patientList)
+	if err != nil {
+		return nil, err
+	}
+
+	return c.getPayments(patientID)
+}
+
+func (c *Client) getPayments(patient int) (map[string]interface{}, error) {
+
+	url := fmt.Sprintf("%s/patients/%d/payments", c.url, patient)
+	log.Printf("Fetching all payments for patient %d", patient)
+	log.Println(url)
+	resp, err := resty.R().
+		Get(url)
+	if err != nil {
+		return nil, err
+	}
+
+	bodyMap, err := parseHTTPResponseReturnMap(resp)
+	if err != nil {
+		return nil, err
+	}
+
+	return bodyMap, err
+}

--- a/go/daf_tests/records.go
+++ b/go/daf_tests/records.go
@@ -1,0 +1,35 @@
+package main
+
+import (
+	"fmt"
+	"log"
+
+	resty "gopkg.in/resty.v1"
+)
+
+func (c *Client) getRecordsFromAnyPatientInList(patientList map[string]interface{}) (map[string]interface{}, error) {
+	patientID, err := getPseudoRandomPatientIDFromPatientsList(patientList)
+	if err != nil {
+		return nil, err
+	}
+	return c.getRecords(patientID)
+}
+
+func (c *Client) getRecords(patient int) (map[string]interface{}, error) {
+
+	url := fmt.Sprintf("%s/patients/%d/records", c.url, patient)
+	log.Printf("Fetching all records for patient %d", patient)
+	log.Println(url)
+	resp, err := resty.R().
+		Get(url)
+	if err != nil {
+		return nil, err
+	}
+
+	bodyMap, err := parseHTTPResponseReturnMap(resp)
+	if err != nil {
+		return nil, err
+	}
+
+	return bodyMap, nil
+}

--- a/go/daf_tests/users.go
+++ b/go/daf_tests/users.go
@@ -1,0 +1,99 @@
+package main
+
+import (
+	"fmt"
+	"log"
+	"net/http"
+
+	resty "gopkg.in/resty.v1"
+)
+
+type User struct {
+	username, firstname, lastname, password string
+}
+
+func (c *Client) getUsers() map[string]interface{} {
+
+	url := c.url + "/users/"
+	log.Println("Fetching all users")
+	log.Println(url)
+	resp, err := resty.R().
+		Get(url)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	if http.StatusUnauthorized == resp.StatusCode() {
+		err = fmt.Errorf("%d", http.StatusUnauthorized)
+		if err != nil {
+			log.Fatal(err)
+		}
+	}
+	bodyMap, err := parseHTTPResponseReturnMap(resp)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	return bodyMap
+}
+
+func (c *Client) signUpUser(user *User) (map[string]interface{}, error) {
+
+	url := c.url + "/auth/sign-up"
+	log.Println(url)
+	log.Printf("Creating %v", user)
+	payload := fmt.Sprintf("{\n\t\"username\": \"%s\",\n\t\"firstname\": \"%s\",\n\t\"lastname\": \"%s\"\n,\n\t\"password\": \"%s\"\n}", user.username, user.firstname, user.lastname, user.password)
+	log.Printf(payload)
+	resp, err := resty.R().
+		SetBody(payload).
+		Post(url)
+	if err != nil {
+		return nil, err
+	}
+
+	if http.StatusUnauthorized == resp.StatusCode() {
+		err = fmt.Errorf("%d", http.StatusUnauthorized)
+		if err != nil {
+			return nil, err
+		}
+	}
+	bodyMap, err := parseHTTPResponseReturnMap(resp)
+	if err != nil {
+		return nil, err
+	}
+
+	return bodyMap, nil
+}
+
+func (c *Client) deleteUser(userId int) error {
+
+	url := fmt.Sprintf("%s/users/%d", c.url, userId)
+	log.Println(url)
+	log.Printf("Deleting %d", userId)
+	resp, err := resty.R().
+		Delete(url)
+	if err != nil {
+		return err
+	}
+
+	if http.StatusUnauthorized == resp.StatusCode() {
+		err = fmt.Errorf("%d", http.StatusUnauthorized)
+		if err != nil {
+			return err
+		}
+	}
+	_, err = parseHTTPResponse(resp)
+	if err != nil {
+		return err
+	}
+	log.Printf("UserID %d deleted", userId)
+	return nil
+}
+
+func (c *Client) signUpUserAndReturnID(user *User) (int, error) {
+	id, err := c.signUpUser(user)
+	if err != nil {
+		return 0, err
+	}
+	return int(id["id"].(float64)), nil
+}

--- a/packer-templates/delphix-tcw-tooling-oracle.json
+++ b/packer-templates/delphix-tcw-tooling-oracle.json
@@ -109,7 +109,7 @@
     "labalchemy_pub_key": "{{env `LABALCHEMY_PUB_KEY`}}",
     "md5sum": "",
     "region": "{{env `AWS_REGION`}}",
-    "seed": "191730",
+    "seed": "1ea60b1",
     "source": "delphix-centos7-tooling-base.json",
     "subnet_id": "{{env `AWS_SUBNET_ID`}}",
     "vpc_id": "{{env `AWS_VPC_ID`}}"

